src/visitor.rs

```diff
@@ -0,0 +1,769 @@
+// Copyright (C) 2024 Parity Technologies (UK) Ltd. (admin@parity.io)
+// This file is a part of the scale-decode crate.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//         http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#![allow(clippy::type_complexity)]
+
+use crate::{
+    BitsOrderFormat, BitsStoreFormat, Field, FieldIter, Primitive, ResolvedTypeVisitor,
+    UnhandledKind, Variant, VariantIter,
+};
+use smallvec::SmallVec;
+
+/// A concrete iterator over struct or variant fields information.
+pub struct ConcreteFieldIter<'resolver, TypeId> {
+    fields: SmallVec<[Field<'resolver, TypeId>; 16]>,
+    idx: usize,
+}
+
+impl<'resolver, TypeId> Iterator for ConcreteFieldIter<'resolver, TypeId> {
+    type Item = Field<'resolver, TypeId>;
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.fields.get(self.idx) {
+            Some(field) => {
+                self.idx += 1;
+                Some(*field)
+            }
+            None => None,
+        }
```

Comment:
Nit: could also be bit shorter:
```suggestion
        let field = self.fields.get(self.idx)?;
        self.idx +=1;
        Some(*field)
```

Reply:
Ooh yup that's nicer!